

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>3. Prepare your raw datafiles &mdash; Stacks workflow tutorial 1.0 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="Stacks workflow tutorial 1.0 documentation" href="index.html"/>
        <link rel="next" title="4. Extract individual data with process_radtags" href="04-Step_4.html"/>
        <link rel="prev" title="2. Installing Stacks_workflow" href="02-Step_2.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> Stacks workflow tutorial</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="01-stacks.html">1. Stacks workflow tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="01-stacks.html#about-stacks">1.1. About STACKS</a></li>
<li class="toctree-l2"><a class="reference internal" href="01-stacks.html#overview-of-the-steps">1.2. Overview of the steps</a></li>
<li class="toctree-l2"><a class="reference internal" href="01-stacks.html#where-to-find-this-manual">1.3. Where to find this manual</a></li>
<li class="toctree-l2"><a class="reference internal" href="01-stacks.html#licence">1.4. Licence</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="02-Step_2.html">2. Installing Stacks_workflow</a><ul>
<li class="toctree-l2"><a class="reference internal" href="02-Step_2.html#download-and-install-the-most-recent-version-of-this-workflow">2.1. Download and install the most recent version of this workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="02-Step_2.html#set-up-for-mac-osx">2.2. Set up for Mac OSX</a></li>
<li class="toctree-l2"><a class="reference internal" href="02-Step_2.html#download-and-install-stacks">2.3. Download and install STACKS</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">3. Prepare your raw datafiles</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#downloading-your-data">3.1. Downloading your data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#copy-your-raw-files">3.2. Copy your raw files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preparing-the-lane-info-txt-file">3.3. Preparing the <tt class="docutils literal"><span class="pre">lane_info.txt</span></tt> file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#removing-adapters">3.4. Removing adapters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#assess-sequence-quality-with-fastqc">3.5. Assess sequence quality with FastQC</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="04-Step_4.html">4. Extract individual data with <tt class="docutils literal"><span class="pre">process_radtags</span></tt></a><ul>
<li class="toctree-l2"><a class="reference internal" href="04-Step_4.html#prepare-a-file-named-sample-information-csv">4.1. Prepare a file named <tt class="docutils literal"><span class="pre">sample_information.csv</span></tt></a></li>
<li class="toctree-l2"><a class="reference internal" href="04-Step_4.html#launch-process-radtags-with">4.2. Launch process_radtags with:</a></li>
<li class="toctree-l2"><a class="reference internal" href="04-Step_4.html#testing-different-trim-lengths">4.3. Testing different trim lengths</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="05-Step_5.html">5. Rename samples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="05-Step_5.html#rename-and-copy-the-samples">5.1. Rename and copy the samples</a></li>
<li class="toctree-l2"><a class="reference internal" href="05-Step_5.html#deleting-samples-with-too-few-reads">5.2. Deleting samples with too few reads</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="06-Step_6.html">6. Align reads to a reference genome (optional)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="06-Step_6.html#install-bwa">6.1. Install bwa</a></li>
<li class="toctree-l2"><a class="reference internal" href="06-Step_6.html#download-reference-genome-to-the-01-info-files">6.2. Download reference genome to the <tt class="docutils literal"><span class="pre">01-info_files</span></tt></a></li>
<li class="toctree-l2"><a class="reference internal" href="06-Step_6.html#index-the-reference-genome">6.3. Index the reference genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="06-Step_6.html#align-samples">6.4. Align samples</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="07-Step_7.html">7. STACKS pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="07-Step_7.html#prepare-population-info-file">7.1. Prepare population info file</a></li>
<li class="toctree-l2"><a class="reference internal" href="07-Step_7.html#edit-script-parameters">7.2. Edit script parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="07-Step_7.html#run-the-stacks-programs">7.3. Run the STACKS programs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="08-Step_8.html">8. Filtering the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="09-conclusion.html">9. Conclusion</a><ul>
<li class="toctree-l2"><a class="reference internal" href="09-conclusion.html#running-into-problems">9.1. Running into problems</a></li>
</ul>
</li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Stacks workflow tutorial</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>3. Prepare your raw datafiles</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/03-Step_3.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <div class="section" id="prepare-your-raw-datafiles">
<h1>3. Prepare your raw datafiles<a class="headerlink" href="#prepare-your-raw-datafiles" title="Permalink to this headline">¶</a></h1>
<div class="section" id="downloading-your-data">
<h2>3.1. Downloading your data<a class="headerlink" href="#downloading-your-data" title="Permalink to this headline">¶</a></h2>
<p>Download your raw Illumina or Ion Proton data files from your sequencing
service provider.</p>
</div>
<div class="section" id="copy-your-raw-files">
<h2>3.2. Copy your raw files<a class="headerlink" href="#copy-your-raw-files" title="Permalink to this headline">¶</a></h2>
<p>Put a copy of (or a link to) your raw data files in the <tt class="docutils literal"><span class="pre">02-raw</span></tt> folder of
stacks_workflow.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">All file names <strong>must</strong> end with <strong>.fastq.gz</strong> for the following scripts to
work.</p>
</div>
</div>
<div class="section" id="preparing-the-lane-info-txt-file">
<h2>3.3. Preparing the <tt class="docutils literal"><span class="pre">lane_info.txt</span></tt> file<a class="headerlink" href="#preparing-the-lane-info-txt-file" title="Permalink to this headline">¶</a></h2>
<p>This file will contains the names of the raw data files and is used by
stacks_workflow later.  From the stacks_workflow folder, run:</p>
<div class="highlight-bash"><div class="highlight"><pre>./00-scripts/00_prepare_lane_info.sh
</pre></div>
</div>
</div>
<div class="section" id="removing-adapters">
<h2>3.4. Removing adapters<a class="headerlink" href="#removing-adapters" title="Permalink to this headline">¶</a></h2>
<p>We use Cutadapt in order to remove the adapters present in the raw data. This
step is done before we extract the reads for each individual.</p>
<div class="section" id="creating-a-file-with-adapters">
<h3>3.4.1. Creating a file with adapters<a class="headerlink" href="#creating-a-file-with-adapters" title="Permalink to this headline">¶</a></h3>
<p>In order to clean your reads properly, you must determine which adapters are
present in your data and whether a cut site can also be found with it.</p>
<p>Before running cutadapt, we must create a fasta file containing the adapters
that are present in our data. This file must be found in the <tt class="docutils literal"><span class="pre">01-info_files/</span></tt>
folder and be names <tt class="docutils literal"><span class="pre">adapters.fasta</span></tt>. The file named
<tt class="docutils literal"><span class="pre">example_adapters.fasta</span></tt>, found in the <tt class="docutils literal"><span class="pre">01-info_files/</span></tt> folder should be
used as a template.</p>
<p>The best way to know which adapters and potential restriction enzyme cut sites
are present in your data is probably to ask the people who created your
sequencing libraries.</p>
<p>In order to identify or confirm the potential adapters found in your data
directly, you can use the following command (you will need to have installed
<tt class="docutils literal"><span class="pre">jellyfish</span></tt>):</p>
<div class="highlight-bash"><div class="highlight"><pre>cat <span class="k">$(</span>ls -1 02-raw/*.fastq.gz | head -1<span class="k">)</span> | <span class="se">\</span>
   gunzip -c | grep -E <span class="s2">&quot;^[ACTG]+$&quot;</span> | head -10000 | cut -c 20- | <span class="se">\</span>
   awk <span class="s1">&#39;length &gt; 30 {print &quot;&gt;seq_&quot; NR &quot;\n&quot; $_}&#39;</span> &gt; temp.fasta

rm kmer.count.*
<span class="k">for </span>k in <span class="k">$(</span>seq 12 30<span class="k">)</span>
<span class="k">do</span>
<span class="k">   </span>jellyfish count -m <span class="nv">$k</span> -s 10000 -t 8 --text -o kmer.count.<span class="s2">&quot;$k&quot;</span> temp.fasta
   perl -pe <span class="s1">&#39;s/^.+\}[^ACTG]+//&#39;</span> kmer.count.<span class="s2">&quot;$k&quot;</span> | awk <span class="s1">&#39;{print $2,$1}&#39;</span> | <span class="se">\</span>
   sort -nr | head -20 &gt; kmer.count.sorted.<span class="s2">&quot;$k&quot;</span>; <span class="k">done</span>
</pre></div>
</div>
<p>Then look at the results with the following command. From the number of
identical kmers and how they vary from one kmer length to the next, you should
be able to spot your adapters and potentially a version containing a cut site
just before it if you used two restriction enzymes in your library preparation.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="k">for </span>i in kmer.count.sorted.*; <span class="k">do </span><span class="nb">echo</span> <span class="nv">$i</span>; cat <span class="nv">$i</span>; <span class="k">done</span> | less

<span class="c"># Delete all these files</span>
rm kmer.count*
</pre></div>
</div>
</div>
<div class="section" id="running-cutadapt">
<h3>3.4.2. Running Cutadapt<a class="headerlink" href="#running-cutadapt" title="Permalink to this headline">¶</a></h3>
<p>Now that the <tt class="docutils literal"><span class="pre">01-info_files/adapters.fasta</span></tt> file contains all the adapters we
wish to remove from our data, we can launch Cutadapt <strong>in single-end mode</strong>
with the following command:</p>
<div class="highlight-bash"><div class="highlight"><pre>./00-scripts/01_cutadapt.sh
</pre></div>
</div>
</div>
<div class="section" id="scan-the-cutadapt-logs">
<h3>3.4.3. Scan the cutadapt logs<a class="headerlink" href="#scan-the-cutadapt-logs" title="Permalink to this headline">¶</a></h3>
<p>The cutadapt log files can be found in the <tt class="docutils literal"><span class="pre">98-log_files</span></tt> folder. Scan them
to confirm that cutadapt has done an appropriate job.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Ther may be difference in adapters and filter parameters to use with data
produced by Illumina and Ion Proton sequencers.</p>
</div>
</div>
</div>
<div class="section" id="assess-sequence-quality-with-fastqc">
<h2>3.5. Assess sequence quality with FastQC<a class="headerlink" href="#assess-sequence-quality-with-fastqc" title="Permalink to this headline">¶</a></h2>
<p><strong>TODO:</strong></p>
<ul class="simple">
<li>Before and after Cutadapt</li>
<li>Use the script
- Command line interface
- Run only on first 1,000,000 reads or less
- Gathers important figures together</li>
</ul>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="04-Step_4.html" class="btn btn-neutral float-right" title="4. Extract individual data with process_radtags"/>Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="02-Step_2.html" class="btn btn-neutral" title="2. Installing Stacks_workflow"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
      Last updated on Jul 14, 2015.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>